### PR TITLE
Clarity get/set! when canget/set returns false

### PR DESCRIPTION
**Pull request by odow**
As raised in https://github.com/JuliaOpt/MathOptInterface.jl/issues/209

### CODE REVIEW

**Comment by mlubin** [comment]
Hmm. I don't think we should allow `get` to return incorrect results or say that `set!` values can be ignored. These should be errors.

**Comment by tkoolen** [comment]
I think that ideally:

* It should always be possible to get any attribute in a 'safe' way, e.g. by having a `checkedget`/`safeget` (or maybe *this* should be called `get`) that first checks `canget` and then returns the result. This should probably be the default behavior.
* There should be an `unsafeget` version (currently called `get`) for at least some properties that foregoes the `canget` check for efficiency reasons, (e.g when you want to iterate over a bunch of `VariableIndex`es) and you've already checked `canget` for `::Type{VariableIndex}`.

This is starting to sound like the `@inbounds` machinery. Unfortunately, that's hard to replicate, but maybe some kind of simple `@unsafe` macro solution to redirect to the `unsafeget` version would be nice.

**Comment by mlubin** [comment]
@tkoolen is there really a need for an unsafe version? If you're iterating over a bunch of `VariableIndex`es then you can use the ``::Vector{VariableIndex}`` version of `get` instead.

**Comment by blegat** [comment]
What would be done in `unsafeget` if `canget` is `false` ? The solvers returns an incorrect value and the error is thrown in `safeget` which is implemented in MOI like this ?
```julia
function MOI.safeget(...)
    if MOI.canget(...)
        MOI.get(...)
    else
        error(...)
    end
end
```

**Comment by tkoolen** [comment]
@mlubin, I thought that there was a need for that given your earlier comment, https://github.com/JuliaOpt/MathOptInterfaceUtilities.jl/issues/87#issuecomment-364540567. If that's not the case, what I proposed is probably unnecessary.

@blegat, yep.

**Comment by odow** [comment]
This has been updated to reflect that the call should throw an error.

File: src/attributes.jl [hunk 1]
```diff
@@ -96,6 +96,11 @@ If an `F`-in-`S` constraint with name `name` exists in the model `model`, return
 
 If *any* constraint with name `name` exists in the model `model`, return the corresponding index, otherwise throw a `KeyError`. This version is available for convenience but may incur a performance penalty because it is not type stable.
 
+### Note
+
+It is the user's responsibility to check that `canget` returns `true` prior to calling `get`. 
+Calling `get` when `canget` returns `false` will throw an error.
```
Comment (mlubin) [inline src/attributes.jl]: This is now a weird statement. Do we need to add this note for `addconstraint!` `modifyconstraint!`, etc (anything that as a corresponding `canXXX`)?
I would phrase it more of a note for implementations. These functions *must* throw errors if the corresponding `canXXX` returned false.

Comment (odow) [reply src/attributes.jl]: Should this just be a general note in the manual then?

Comment (mlubin) [reply src/attributes.jl]: Yeah, could be in the section on notes for solver wrappers.

**Comment by odow** [comment]
Closing in favour of #228

**Pull request closed without merging**